### ISSUE TITLE
[Go Function API] Ability to store state

### ISSUE THREAD

**Issue by mrijkeboer** [apache/pulsar#5138]
Currently the Go Function API doesn't seem to have the ability to store state like the Java Function API. It would be useful if this functionality would be available to Go as well.

**Comment by phemmer** [comment]
I've created a wrapper around `pulsar-function-go/pf` which provides state support. It's pretty minimal at the moment, and needs more work (concurrency safety & documentation), but it does function. If you're interested, it can be found here: https://godoc.org/github.com/phemmer/go-pulsar-state/pulsar

An example of it's usage can be found here: https://github.com/phemmer/go-pulsar-state/blob/master/pulsar/examples/wordcount/wordcount.go

**Comment by phemmer** [comment]
This one has more labels, more attention from devs, and is older. I'd suggest keeping this one.

**Comment by tisonkun** [comment]
Closed as stale. Please open a new issue if it's still relevant to the maintained versions.

**Comment by flowchartsman** [comment]
Considering as the function api for go does not, and has never had this
functionality, I’d say it’s still relevant, yeah.

**Comment by tisonkun** [comment]
@flowchartsman Good to know. A wishing list or open-ended discussion can be started at https://github.com/apache/pulsar/discussions/categories/general. We use issues to track actionable tasks, so only with a tech design or a volunteer to work on it, it's valuable to keep it open.

Open-source projects evolve with contributions, not requirements. Open such an issue for 4 years cannot implement it automatically :)

**Comment by flowchartsman** [comment]
> Open-source projects evolve with contributions, not requirements. Open such an issue for 4 years cannot implement it automatically :)

Okay, so is the push to move things to discussion and then to have a "feature" issue filed only once people have agreed on the implementation and to work on it? I can get behind that.

**Comment by flowchartsman** [comment]
Understood, but I don't think this issue can be written off simply as a wishlist item that never went anywhere.  This issue appeared to have support and partial implementations, but was just waiting on a maintainer to come and finish it.  In particular, it looked as if it was getting worked on by a project member already with the [PR](https://github.com/apache/bookkeeper/pull/2750 ) to add Go bookeeper support mentioned on the [other issue I filed](https://github.com/apache/pulsar/issues/10560), and that PR is still sitting open with no attention. What's the status of it? What needs to happen? Is it going to be closed as stale too? If so, is it our responsibility to create a new discussion for it so that the existing work and context is not lost?

It's not as if the people commenting on this issue _can't_ do the work, but we've just kind of been waiting on more movement from the work that already appeared underway.

**Comment by tisonkun** [comment]
@flowchartsman That sounds good :) Thanks for nudging the initiative!

I'll then lock this conversation to avoid fragmented discussions. You're welcome to open new issues or call for reviews of existing works - rolling up legacy tickets can improve their visibility, and most of them just fixed among the years or no longer relevant.
